# Predicting Hotel Booking Cancellations with PyTorch: A Machine Learning Approach

## 📚 Project Overview

This project focuses on using **PyTorch** and **neural networks** to predict hotel booking cancellations, leveraging real-world data from a resort hotel. The goal is to build robust machine learning models that help hotels make data-driven decisions to optimize their operations and revenue strategies.

## 🎯 Project Goals

The main objectives of this project are:

1. **Develop predictive models using PyTorch:**
   - A **binary classification model** to predict whether a customer will cancel their booking or not.
   - A **multiclass classification model** to predict whether a customer will:
     - Show up to their booking
     - Cancel their booking in advance
     - No-show without prior notice

2. **Enhance hotel management strategies:**
   - **Revenue optimization:** Predicting cancellations allows hotels to better manage overbookings and set dynamic pricing.
   - **Resource allocation:** Accurate predictions help optimize staff and amenity management.
   - **Marketing strategies:** Identify high-risk customers and tailor ad campaigns or special offers to reduce cancellations.

## 📊 Dataset

The dataset used for this project contains comprehensive booking data, with features that provide critical insights into customer behavior. 

**Key features include:**
- **Booking date:** When the reservation was made.
- **Length of stay:** Number of nights the customer plans to stay.
- **Guests:** Number of adults, children, and babies per booking.
- **Daily rates:** Average price per night.
- **Reservation status:** Indicates whether the customer checked in, canceled in advance, or no-showed.

You can access the dataset [here](https://www.kaggle.com/datasets/jessemostipak/hotel-booking-demand).

## ⚙️ Workflow

The project workflow follows a structured machine learning pipeline:

1. **Data Exploration and Cleaning:**
   - Handle missing values and outliers.
   - Convert categorical data into numerical representations.
   - Feature engineering to enhance predictive power.

2. **Data Preprocessing:**
   - Split data into training, validation, and test sets.
   - Normalize numerical features.
   - Encode target variables for binary and multiclass models.

3. **Model Building:**
   - Design two neural network architectures:
     - **Binary classifier:** Uses a sigmoid activation function for the final output layer.
     - **Multiclass classifier:** Implements softmax activation for multi-category prediction.
   - Define loss functions (Binary Cross-Entropy and Categorical Cross-Entropy).

4. **Training and Evaluation:**
   - Use appropriate optimizers (Adam, SGD).
   - Implement early stopping and learning rate scheduling.
   - Evaluate models using metrics such as accuracy, precision, recall, and F1 score.

5. **Model Interpretation and Insights:**
   - Visualize feature importance.
   - Create confusion matrices to understand model errors.
   - Extract actionable insights for hotel management.

## 📈 Conclusion

By the end of this project, the models will provide actionable predictions for hotel booking behaviors. These insights will help hotels reduce financial loss due to cancellations, improve customer targeting strategies, and enhance overall operational efficiency.

---

**Setup - Import libraries**

In [ ]:
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

## Import and Inspect

The file `'datasets/resort_hotel_bookings.csv'` contains a subset of a [real-world dataset](https://www.kaggle.com/datasets/jessemostipak/hotel-booking-demand) containing reservation and cancellation data for a resort hotel. 

Your goal in this project is build and train a neural network to predict if a customer will cancel their hotel booking reservation based on data including the booking dates, average daily cost, number of adults/children/babies, duration of stay, and so forth.

In [33]:
hotels = pd.read_csv('datasets/resort_hotel_bookings.csv')

Begin by importing the CSV file to a pandas DataFrame named `hotels`.

Preview the first five rows using the `.head()` method.

In [34]:
display(hotels)
print(hotels.head())

is_canceled  lead_time  arrival_date_year arrival_date_month  \
0                0        342               2015               July   
1                0        737               2015               July   
2                0          7               2015               July   
3                0         13               2015               July   
4                0         14               2015               July   
...            ...        ...                ...                ...   
40055            0        212               2017             August   
40056            0        169               2017             August   
40057            0        204               2017             August   
40058            0        211               2017             August   
40059            0        161               2017             August   

       arrival_date_week_number  arrival_date_day_of_month  \
0                            27                          1   
1                            27                          1   
2                            27                          1   
3                            27                          1   
4                            27                          1   
...                         ...                        ...   
40055                        35                         31   
40056                        35                         30   
40057                        35                         29   
40058                        35                         31   
40059                        35                         31   

       stays_in_weekend_nights  stays_in_week_nights  adults  children  ...  \
0                            0                     0       2       0.0  ...   
1                            0                     0       2       0.0  ...   
2                            0                     1       1       0.0  ...   
3                            0                     1       1       0.0  ...   
4                            0                     2       2       0.0  ...   
...                        ...                   ...     ...       ...  ...   
40055                        2                     8       2       1.0  ...   
40056                        2                     9       2       0.0  ...   
40057                        4                    10       2       0.0  ...   
40058                        4                    10       2       0.0  ...   
40059                        4                    10       2       0.0  ...   

       deposit_type  agent company days_in_waiting_list    customer_type  \
0        No Deposit    NaN     NaN                    0        Transient   
1        No Deposit    NaN     NaN                    0        Transient   
2        No Deposit    NaN     NaN                    0        Transient   
3        No Deposit  304.0     NaN                    0        Transient   
4        No Deposit  240.0     NaN                    0        Transient   
...             ...    ...     ...                  ...              ...   
40055    No Deposit  143.0     NaN                    0        Transient   
40056    No Deposit  250.0     NaN                    0  Transient-Party   
40057    No Deposit  250.0     NaN                    0        Transient   
40058    No Deposit   40.0     NaN                    0         Contract   
40059    No Deposit   69.0     NaN                    0        Transient   

          adr  required_car_parking_spaces  total_of_special_requests  \
0        0.00                            0                          0   
1        0.00                            0                          0   
2       75.00                            0                          0   
3       75.00                            0                          0   
4       98.00                            0                          1   
...       ...                          ...                        ...   
40055   89.75                       

   is_canceled  lead_time  arrival_date_year arrival_date_month  \
0            0        342               2015               July   
1            0        737               2015               July   
2            0          7               2015               July   
3            0         13               2015               July   
4            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  ...  \
0                        0                     0       2       0.0  ...   
1                        0                     0       2       0.0  ...   
2                      

<details><summary style="display:list-item; font-size:16px; color:blue;">Here's a quick summary of the columns</summary>

- **is_canceled**: Whether the booking was canceled (1) or kept (0)
- **lead_time**: Number of days between booking date and arrival date
- **arrival_date_year**: Year of arrival date
- **arrival_date_month**: Month of arrival date
- **arrival_date_week_number**: Week number of arrival date
- **arrival_date_day_of_month**: Day of the month of arrival date
- **stays_in_weekend_nights**: Number of weekend nights booked (Sat-Sun)
- **stay_in_week_nights**: Number of weekday nights booked (Mon-Fri)
- **adults**: Number of adults
- **children**: Number of children
- **babies**: Number of babies
- **meal**: Type of meal booked (Undefined/SC, BB, HB, or FB)
- **country**: Country of origin of the booker
- **market_segment**: Market segment (TA - travel agent, TO - tour operators)
- **distribution_channel**: Booking distribution channel (TA - travel agent, TO - tour operators)
- **is_repeated_guest**: Is this a repeated guest (1) or not (0)
- **previous_cancellations**: The number of previous bookings canceled by the customer
- **previous_bookings_not_canceled**: The number of previous bookings not canceled by the customer
- **reserved_room_type**: Room type reserved
- **assigned_room_type**: Type of assigned room booked
- **booking_changes**: Number of booking changes or modifications
- **deposit_type**: Type of deposit to guarantee booking (No Deposit, Non Refund, or Refundable)
- **agent**: ID of the travel agency that made the booking
- **company**: ID of the company that made the booking
- **days_in_waiting_list**: Number of days booking was waitlisted before confirmation
- **customer_type**: The customer type of booking (Contract, Group, Transient, or Transient-party)
- **adr**: The average daily rate (cost) of the booking
- **required_car_parking_spaces**: Number of parking spaces requested by the customer
- **total_of_special_requests**: Number of special requests by the customer
- **reservation_status**: The last reservation status (Canceled, Check-Out, No-Show)
- **reservation_status_date**: The date of the last reservation status

Let's explore the data types and whether any data is missing.

Use the `.info()` method on the `hotels` DataFrame to inspect the data.

In [6]:
hotels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40060 entries, 0 to 40059
Data columns (total 31 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   is_canceled                     40060 non-null  int64  
 1   lead_time                       40060 non-null  int64  
 2   arrival_date_year               40060 non-null  int64  
 3   arrival_date_month              40060 non-null  object 
 4   arrival_date_week_number        40060 non-null  int64  
 5   arrival_date_day_of_month       40060 non-null  int64  
 6   stays_in_weekend_nights         40060 non-null  int64  
 7   stays_in_week_nights            40060 non-null  int64  
 8   adults                          40060 non-null  int64  
 9   children                        40060 non-null  float64
 10  babies                          40060 non-null  int64  
 11  meal                            40060 non-null  object 
 12  country                         

<details><summary style="display:list-item; font-size:16px; color:blue;">What do we notice about the dataset under inspection?</summary>

There are 31 columns and 40,060 total observations in our dataset. The majority of columns do not have missing values.

However, we do notice that: 
- the `agent` and `company` columns seem to have missing values that need to be addressed
- the `country` column has a couple of missing values as well

There are a variety of data types represented. To work with a neural network, we'll have to address any non-numeric columns in our data preparation.

Let's now explore the cancellation column we want to predict.

Use the `.value_counts()` method on the `is_canceled` column to count the number **and** the percentage of overall cancellations. 

In [7]:
# total number of cancelations
hotels.is_canceled.value_counts()

is_canceled
0    28938
1    11122
Name: count, dtype: int64

The `reservation_status` column tells us if the booking was canceled while also telling us if the customer was a no-show.

We need to be sure to exclude this column from the training set, otherwise this information will be _leaked_ to our model resulting in inaccurate performance. 

First, let's take a quick look at the values in this column.

Use the `.value_counts()` method on the `reservation_status` column to count the number **and** the percentage of overall cancellations. 

In [8]:
# Total Number of cancellations
print(hotels['is_canceled'].value_counts(0))

# Percentage of cancellations
print(hotels['is_canceled'].value_counts(1))

is_canceled
0    28938
1    11122
Name: count, dtype: int64
is_canceled
0    0.722366
1    0.277634
Name: proportion, dtype: float64


Before diving into building a model, let's continue to explore the dataset. It's important to understand how different columns interact with cancellations to guide our model structure! 

For example, cancellations might be higher in the summer months (June - September) and lower in the winter months (November - January).

Use the `.groupby()` method to group the data by the `arrival_date_month` column and apply the `.mean()` aggregation function on the `is_canceled` column. This will return the percent of reservations cancelled in each month.

Then, use the `.sort_values()` method to sort the percentages from lowest to highest.

In [9]:
grouped_data = hotels.groupby('arrival_date_month')['is_canceled'].mean()
grouped_data.sort_values()

arrival_date_month
January      0.148199
November     0.189167
March        0.228717
December     0.238293
February     0.256204
October      0.275105
May          0.287721
April        0.293433
July         0.314017
September    0.323681
June         0.330706
August       0.334491
Name: is_canceled, dtype: float64

<details><summary style="display:list-item; font-size:16px; color:blue;">What do we notice about the percentage of cancellations by month?</summary>

Winter and spring have the lowest cancellation percentages, while summer and fall have the highest. This information can be very useful for our model!

It might be useful to do more exploratory data analysis to gain additional insights about hotel cancellations. For example, additional analysis may help you select better features to train the model on and exclude features that might seem irrelevant. But for now, let's move on to cleaning and preparing the data.

## Data Cleaning and Preparation

In this section, we'll encode categorical data for use in our neural networks.

To get a sense of the categorical data in the dataset, let's start by previewing the first five rows of all columns with `object` datatype.

Create a list named `object_columns` containing only the names of the object columns (except for the reservation status columns). Select those columns from `hotels` and preview the first `5` rows.

In [10]:
object_columns = ['arrival_date_month', 'meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'customer_type']
hotels[object_columns].head()

# Seleccionar solo las columnas de tipo object excepto 'reservation_status' que la voy a excluir
# object_columns = hotels.select_dtypes(include='object').drop(columns=['reservation_status'])
# object_columns.head()

arrival_date_month meal country market_segment distribution_channel  \
0               July   BB     PRT         Direct               Direct   
1               July   BB     PRT         Direct               Direct   
2               July   BB     GBR         Direct               Direct   
3               July   BB     GBR      Corporate            Corporate   
4               July   BB     GBR      Online TA                TA/TO   

  reserved_room_type assigned_room_type deposit_type customer_type  
0                  C                  C   No Deposit     Transient  
1                  C                  C   No Deposit     Transient  
2                  A                  C   No Deposit     Transient  
3                  A                  A   No Deposit     Transient  
4                  A                  A   No Deposit     Transient

Typically, we don't want to use every column in training. For example, we may want to drop columns with many missing values or columns that are irrelevant to our prediction task.

Drop any columns you don't want to use to train a cancellation model (do not remove the target label column). Feel free to open our Hint to review the columns we chose to drop in our solution.

Note: We don't want to drop the `reservation_status` column from the dataset quite yet because we'll be using this column to train our multiclass neural network.

In [11]:
columns_to_drop = ['country', 'agent', 'company', 'reservation_status_date', 'arrival_date_week_number', 'arrival_date_day_of_month', 'arrival_date_year']

hotels = hotels.drop(labels=columns_to_drop, axis=1)

<details><summary style="display:list-item; font-size:16px; color:blue;">Hint: Drop columns in the dataset not used for training.</summary>

Here's a list of potential features to drop. Feel free to experiment on your own by dropping or keeping columns we might believe may contribute to training.

Here's why we chose these columns:

- `country` - there are many countries that only appear a handful of times in the dataset which may make our model less generalizable and even discriminate against customers based on their country
- `agent` - similar to `country`, there are many agents that only appear a handful of times which may make our model less generalizable (and there are many missing values!)
- `company` - similar to `agent`, there are many companies that only appear a handful of times which may make our model less generalizable (and there are many missing values!)
- `reservation_status_date` - tells us the date of the latest status change of the reservation which shouldn't be helpful and if anything may leak data
- `arrival_date_week_number` - tells us the week of the year which may be too specific and prone to overfitting
- `arrival_date_day_of_month` - tells us the day of the month which may be too specific and prone to overfitting
- `arrival_date_year` - tells us the year of the booking which may not be helpful to predict future years

</details>

Next, let's encode the `meal` column which tells us which type of meal(s) the customer booked: 

- `Undefined` and `SC` correspond to no meal packages
- `BB` corresponds to breakfast only
- `HB` (half board) corresponds to breakfast + lunch or dinner
- `FB` (full board) corresponds to breakfast, lunch, and dinner.

Label encode the `meal` column with a meaningful order (# of meals booked) using the following scheme:

- `Undefined` and `SC` to `0`
- `BB` to `1`
- `HB` to `2`
- `FB` to `3` 

In [12]:
# label encoding. Replace values in 'meal' column
hotels['meal'] = hotels['meal'].replace({'Undefined':0, 'SC':0, 'BB':1, 'HB':2, 'FB':3})

Let's prepare the rest of the categorical columns using one-hot encoding. 

Create a list named `one_hot_columns` containing the list of categorical column names (all the remaining categorical columns) to be one-hot encoded using the `pd.get_dummies()` method.

Preview the cleaned `hotels` DataFrame using the `.head()` method.

In [13]:
one_hot_columns = ['arrival_date_month', 'distribution_channel', 'reserved_room_type', 
                   'assigned_room_type', 'deposit_type', 'customer_type', 'market_segment']

# Verifica las columnas que existen en el DataFrame
existing_columns = [col for col in one_hot_columns if col in hotels.columns]

# Aplica el one-hot encoding solo a las columnas que existen
hotels = pd.get_dummies(hotels, columns=existing_columns, dtype=int)

hotels.head()
hotels.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40060 entries, 0 to 40059
Data columns (total 67 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   is_canceled                     40060 non-null  int64  
 1   lead_time                       40060 non-null  int64  
 2   stays_in_weekend_nights         40060 non-null  int64  
 3   stays_in_week_nights            40060 non-null  int64  
 4   adults                          40060 non-null  int64  
 5   children                        40060 non-null  float64
 6   babies                          40060 non-null  int64  
 7   meal                            40060 non-null  int64  
 8   is_repeated_guest               40060 non-null  int64  
 9   previous_cancellations          40060 non-null  int64  
 10  previous_bookings_not_canceled  40060 non-null  int64  
 11  booking_changes                 40060 non-null  int64  
 12  days_in_waiting_list            

Perfect! It looks like we've handled all of the categorical variables and prepared the DataFrame for training.

Note that the cleaned DataFrame now has 67 columns due to the additional columns created using one-hot encoding.

## Create Training and Testing Sets

Next, let's convert our dataset into PyTorch tensors and split them into training and testing sets.

In [14]:
import torch
import torch.nn as nn
import torch.optim as optim

We need to start by separating our training features from the target labels.

Create a list named `train_features` that contains all of the feature names (column names excluding the target variables `is_canceled` and `reservation_status`).

In [16]:
# Prepare features and target for binary classification
features_binary = hotels.drop(columns=['is_canceled', 'reservation_status'])
features_binary = features_binary.apply(pd.to_numeric, errors='coerce')  # here we ensure numeric values

Using the list of training features in `train_features`, create `X` and `y` tensors:

- `X` contains the data values from the `train_features` columns
- `y` contains the binary labels in the `is_canceled` column in `hotels`

Both `X` and `y` should have the float datatype.

Be sure to set the correct view of `y` using `.view(-1,1)`

In [17]:
X = torch.tensor(features_binary.values, dtype=torch.float32)
y = torch.tensor(hotels['is_canceled'].values, dtype=torch.float32).view(-1, 1)

Let's now split our data contained in `X` and `y` into training and testing sets.

Import the `train_test_split` module from Scikit-learn's `sklearn.model_selection` library.

Split `X` and `y` using the following scheme:
- Use 80% of the data for the training set `X_train` and `y_train`
- Use 20% of the data for the testing set `X_test` and `y_test`
- Set the random state to `42` to match our solution

We then can print out the shape of `X_train` and `X_test` to see how many observations and columns are in the training and testing sets.

In [18]:
# Split the data
X_train_b, X_test_b, y_train_b, y_test_b = train_test_split(X, y, train_size=0.80, test_size=0.20, random_state=42)

print("Binary Classification - Training Shape:", X_train_b.shape)
print("Binary Classification - Testing Shape:", X_test_b.shape)

Binary Classification - Training Shape: torch.Size([32048, 65])
Binary Classification - Testing Shape: torch.Size([8012, 65])


## Train a Neural Network for Binary Classification

Let's now create a neural network for binary classification to predict hotel cancellations.

Set a random seed to `42` using `torch.manual_seed(42)`.

Build the neural network architecture using `nn.Sequential` with the following:
- input layer with `65` nodes (equal to the number of training features)
- first hidden layer with `36` nodes and a ReLU activation
- second hidden layer with `18` nodes and a ReLU activation
- output layer with `1` node and a Sigmoid activation

In [19]:
torch.manual_seed(42)

# Define the binary classification model
input_dim_b = X_train_b.shape[1]
binary_model = nn.Sequential(
    nn.Linear(input_dim_b, 36),
    nn.ReLU(),
    nn.Linear(36, 18),
    nn.ReLU(),
    nn.Linear(18, 1),
    nn.Sigmoid()
)

Next, let's define the loss function and optimizer used for training:
- set the **binary cross-entropy** loss function to the variable `loss`
- set the **Adam** optimizer to the variable `optimizer` with a learning rate of `0.005`

In [20]:
loss = nn.BCELoss()
optimizer = optim.Adam(binary_model.parameters(), lr=0.005)

Let's build the training loop to train our neural network.

Train the neural network for `1000` epochs.

Keep track of the training performance by printing out the binary cross-entropy loss and accuracy score every `100` epochs.

Before calculating accuracy, convert the model's predicted probabilities to binary labels (as integers) using `0.5` as the threshold.

In [21]:
num_epochs = 1000

for epoch in range(num_epochs):
    predictions = binary_model(X_train_b)
    BCELoss = loss(predictions, y_train_b)
    BCELoss.backward()
    optimizer.step()
    optimizer.zero_grad()
    
    if (epoch + 1) % 100 == 0:
        predicted_labels = (predictions >= 0.5).int()
        accuracy = accuracy_score(y_train_b, predicted_labels)
        print(f'Epoch [{epoch+1}/{num_epochs}], BCELoss: {BCELoss.item():.4f}, Accuracy: {accuracy.item():.4f}')

Epoch [100/1000], BCELoss: 0.3975, Accuracy: 0.8224
Epoch [200/1000], BCELoss: 0.3625, Accuracy: 0.8297
Epoch [300/1000], BCELoss: 0.3537, Accuracy: 0.8347
Epoch [400/1000], BCELoss: 0.3442, Accuracy: 0.8377
Epoch [500/1000], BCELoss: 0.3399, Accuracy: 0.8395
Epoch [600/1000], BCELoss: 0.3339, Accuracy: 0.8409
Epoch [700/1000], BCELoss: 0.3353, Accuracy: 0.8426
Epoch [800/1000], BCELoss: 0.3315, Accuracy: 0.8414
Epoch [900/1000], BCELoss: 0.3263, Accuracy: 0.8450
Epoch [1000/1000], BCELoss: 0.3247, Accuracy: 0.8445


Let's evaluate the trained neural network on the testing set:

1. Set the model to **evaluation mode**
2. Turn off gradient calculations
3. Generate predicted probabilities on `X_test`. Save the probabilities to the variable `test_predictions`.
4. Convert the predicted probabilities to binary labels using `0.5` as the threshold. Save the labels to the variable `test_predicted_labels`.

In [22]:
# Evaluate on test set for binary classification
binary_model.eval()
with torch.no_grad():
    test_predictions = binary_model(X_test_b)
    test_predicted_labels = (test_predictions >= 0.5).int()

Recall that the number of cancellations is much lower than the number of non-cancellations (27.8% canceled vs 72.2% did not cancel). 

To evaluate our neural network effectively, compute the accuracy, precision, recall, and F1 scores using the `sklearn.metrics` module:

- use the `accuracy_score` function to compute the overall accuracy
- use the `classification_report` function to compute the precision, recall, and F1 scores

We will later print out the accuracy and classification report.

In [23]:
from sklearn.metrics import accuracy_score, classification_report

binary_model.eval()
with torch.no_grad():
    test_predictions = binary_model(X_test_b)
    test_predicted_labels = (test_predictions >= 0.5).int()

test_accuracy = accuracy_score(y_test_b.numpy(), test_predicted_labels.numpy())
print(f'Binary Classification Test Accuracy: {test_accuracy:.4f}')
print("Binary Classification Report:")
print(classification_report(y_test_b.numpy(), test_predicted_labels.numpy()))

Binary Classification Test Accuracy: 0.8397
Binary Classification Report:
              precision    recall  f1-score   support

         0.0       0.88      0.90      0.89      5798
         1.0       0.72      0.68      0.70      2214

    accuracy                           0.84      8012
   macro avg       0.80      0.79      0.80      8012
weighted avg       0.84      0.84      0.84      8012



## 📈 Conclusion: Binary Classification Model

The binary classification model demonstrates strong predictive power for identifying hotel booking cancellations:

- **Overall accuracy:** The model achieved an accuracy of **83.7%**, meaning that 83.7% of the predictions were correct.
- **Precision:** When the model predicted a cancellation, it was correct about **72%** of the time.
- **Recall:** The model successfully captured **68%** of all actual cancellations in the data.

While the results are promising, there is room for improvement. A more thorough feature selection process or the inclusion of additional relevant data could enhance predictive performance. For instance, integrating external data like holiday calendars, weather conditions, and economic indicators may boost the model's accuracy.

Additionally, experimenting with the neural network architecture — adjusting the number of nodes, adding hidden layers, using various activation functions, or fine-tuning optimization algorithms — could further refine the model.

Future steps may also include hyperparameter tuning through grid search or random search methods, and implementing cross-validation to ensure the model generalizes well to unseen data.


## Train a Neural Network for Multiclass Classification

Let's now extend our binary classification task to multiclass by attempting to also predict customers who **no-showed** within the `reservation_status` column.

If a hotel can accurately predict no-shows, they can reach out ahead of time to customers who are at high risk of not-showing to their reservation.

First, let's label encode the three categories in the `reservation_status` column:
- **Check-Out** to `2`
- **Canceled** to `1`
- **No-Show** to `0`

In [24]:
#label encoding
hotels['reservation_status'] = hotels['reservation_status'].replace({'Check-Out':2, 'Canceled':1, 'No-Show':0})

Using the same list of training features in `train_features`, create the `X` and `y` tensors where:

- `X` contains the data values from the `train_features` columns
- `y` contains the multiclass data values in the `reservation_status` column

In [26]:
features_multi = hotels.drop(columns=['is_canceled', 'reservation_status'])
features_multi = features_multi.apply(pd.to_numeric, errors='coerce')
X_multi = torch.tensor(features_multi.values, dtype=torch.float32)
y_multi = torch.tensor(hotels['reservation_status'].values, dtype=torch.long)

Similar to before, split the `X` and `y` tensors into training and testing splits using the following scheme:
- Use 80% of the data for the training set `X_train` and `y_train`
- Use 20% of the data for the testing set `X_test` and `y_test`
- Set the random state to `42`

In [27]:
# Split the data
X_train_m, X_test_m, y_train_m, y_test_m = train_test_split(
    X_multi, y_multi, train_size=0.8, test_size=0.2, random_state=42)

print("Multiclass Classification - Training Shape:", X_train_m.shape)
print("Multiclass Classification - Testing Shape:", X_test_m.shape)

Multiclass Classification - Training Shape: torch.Size([32048, 65])
Multiclass Classification - Testing Shape: torch.Size([8012, 65])


Set a random seed using `torch.manual_seed(42)`.

Next, let's construct the multiclass neural network with the following architecture:

- input layer with `65` nodes (equal to the number of training features)
- first hidden layer with `65` nodes and a ReLU activation
- second hidden layer with `36` nodes and a ReLU activation
- final output layer with `3` nodes corresponding to each of the categories in `reservation_status`

We will later save the network to the variable `multiclass_model`.

In [28]:
torch.manual_seed(42)
input_dim_m = X_train_m.shape[1]
multiclass_model = nn.Sequential(
    nn.Linear(input_dim_m, 65),
    nn.ReLU(),
    nn.Linear(65, 36),
    nn.ReLU(),
    nn.Linear(36, 3)  # remember that our output layer should have same number as target classes, in this casses three classes: No-Show (0), Canceled (1), Check-Out (2)
)

Next, let's define the loss function and optimizer used for multiclass training:
- set the **cross-entropy** loss function for multiclass to the variable `loss`
- set the **Adam** optimizer to the variable `optimizer` with a learning rate of `0.01`

In [29]:
criterion_m = nn.CrossEntropyLoss()
optimizer_m = optim.Adam(multiclass_model.parameters(), lr=0.01)

Let's build the training loop to train our neural network.

1. Train the neural network for `500` epochs.
2. Keep track of the training performance by printing out the cross-entropy loss and accuracy score every `100` epochs.
3. Be sure to convert the output probabilites of the multiclass model to labels using the `torch.argmax()` function.

In [30]:
# Training loop for multiclass classification

num_epochs_m = 500
for epoch in range(num_epochs_m):
    multiclass_model.train()
    predictions_m = multiclass_model(X_train_m)
    loss_m = criterion_m(predictions_m, y_train_m)
    loss_m.backward()
    optimizer_m.step()
    optimizer_m.zero_grad()
    
    # Print every 100 epochs
    if (epoch + 1) % 100 == 0:
        predicted_labels_m = torch.argmax(predictions_m, dim=1)
        acc_m = accuracy_score(y_train_m.numpy(), predicted_labels_m.numpy())
        print(f'Epoch [{epoch+1}/{num_epochs_m}], Loss: {loss_m.item():.4f}, Accuracy: {acc_m:.4f}')

Epoch [100/500], Loss: 0.4191, Accuracy: 0.8211
Epoch [200/500], Loss: 0.4071, Accuracy: 0.8217
Epoch [300/500], Loss: 0.3631, Accuracy: 0.8400
Epoch [400/500], Loss: 0.3531, Accuracy: 0.8436
Epoch [500/500], Loss: 0.3459, Accuracy: 0.8475


Let's evaluate the trained neural network on the testing set:

1. Set the multiclass model to **evaluation mode**
2. Turn off gradient calculations
3. Generate predicted probabilities on `X_test`. Save the predicted probabilities to the variable `multiclass_predictions`.
4. Select the class with the largest predicted probability using the `torch.argmax()` function. Save the predicted classes to the variable `multiclass_predicted_labels`.

In [31]:
# Evaluate on test set for multiclass classification
multiclass_model.eval()
with torch.no_grad():
    test_predictions_m = multiclass_model(X_test_m)
    test_predicted_labels_m = torch.argmax(test_predictions_m, dim=1)

Lastly, let's evaluate the multiclass neural network by calculating the overall accuracy, precision, recall, and F1 scores.

Using the `sklearn.metrics` module:
- use the `accuracy_score` function to compute and save the overall accuracy to the variable `multiclass_accuracy`
- use the `classification_report` function to compute and save the classification metrics for each class to the variable `multiclass_report`

Print the overall accuracy and classification report for our multiclass model.

In [32]:
from sklearn.metrics import accuracy_score, classification_report

test_accuracy_m = accuracy_score(y_test_m.numpy(), test_predicted_labels_m.numpy())
print(f'Multiclass Classification Test Accuracy: {test_accuracy_m:.4f}')
print("Multiclass Classification Report:")
print(classification_report(y_test_m.numpy(), test_predicted_labels_m.numpy()))

Multiclass Classification Test Accuracy: 0.8374
Multiclass Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.11      0.19        56
           1       0.72      0.68      0.70      2158
           2       0.88      0.90      0.89      5798

    accuracy                           0.84      8012
   macro avg       0.82      0.56      0.59      8012
weighted avg       0.83      0.84      0.83      8012



## 📈 Conclusion

### Multiclass Neural Network Performance Analysis

Our **multiclass neural network** shows promising results, achieving an **overall accuracy of 84%**, meaning that 84% of all predictions made by the model were correct.

- **Cancellation Predictions:**
  - The **precision score** for predicting cancellations (class 1) is **72%**, indicating that when the model predicts a cancellation, it is correct 72% of the time.
  - The **recall score** is **68%**, meaning the model captures 68% of actual cancellations present in the dataset.

- **No-Show Predictions:**
  - For **no-shows** (class 0), the **precision score** is **86%**, suggesting that when the model predicts a no-show, it is correct 86% of the time — a surprisingly strong result.
  - However, the **recall score** is **11%**, highlighting a significant gap — the model only identifies 11% of actual no-shows, which reduces its effectiveness in capturing these critical cases.
  - The **F1 score** for no-shows is **27%**, pointing to a poor balance between precision and recall, revealing that while the model is confident when predicting a no-show, it misses many actual instances.

### Key Takeaways

- **Strengths:**
  - The model performs well in predicting cancellations, offering hotels a reliable tool to foresee and mitigate revenue loss from advance cancellations.
  - High precision for no-shows means fewer false alarms, helping hotels avoid unnecessary interventions.

- **Weaknesses:**
  - The low recall score for no-shows suggests the model struggles to identify most no-shows, limiting its usefulness for proactive customer engagement.

### Next Steps for Improvement

To enhance performance, future work could include:
- **Data Augmentation:** Collect additional data, such as weather conditions, holiday seasons, and global events like pandemics, to capture more relevant patterns.
- **Feature Engineering:** Explore complex features like last-minute cancellations or booking patterns over time.
- **Model Tuning:** Experiment with:
  - Increasing hidden layers and nodes
  - Trying different activation functions (ReLU, Leaky ReLU)
  - Fine-tuning optimizers and learning rates
  - Incorporating dropout layers to prevent overfitting
- **Class Imbalance Handling:** Use techniques like SMOTE (Synthetic Minority Over-sampling Technique) to balance the dataset, especially for the no-show class.

---